certificates/src/lib.rs

```diff
@@ -108,6 +142,33 @@ impl CertificateValidator {
         // to use with verifying the private key.
         let cert = certs.swap_remove(0);
 
+        if let Some(hostname) = hostname {
+            let dns_names = cert_dns_names(&cert)?;
+            let hostname = DnsName::from_utf8(hostname)
+                .map_err(CertificateError::InvalidDnsName)?;
```

Comment:
This seems like kind of a weird one because the invalid DNS name is not in the certificate, but was provided by the caller, right?

Reply:
Ah, yes, the error message we'd return here would be incorrect. I'll make this its own error case.